src/FuncSharp/Strings/NonEmptyString.cs

```diff
@@ -0,0 +1,61 @@
+﻿using System;
+
+namespace FuncSharp;
+
+public sealed class NonEmptyString : IEquatable<string>
+{
+    private NonEmptyString(string value)
+    {
+        Value = value;
+    }
+
+    public string Value { get; }
+
+    public static implicit operator string(NonEmptyString s)
+    {
+        return s.MapRef(v => v.Value);
+    }
+
+    public static NonEmptyString CreateUnsafe(string value)
+    {
+        var result = CreateNullable(value);
+        if (result is null)
+            throw new ArgumentException("You cannot create NonEmptyString from whitespace, empty string or null.");
+        return result;
+    }
+
+    public static IOption<NonEmptyString> Create(string value)
+    {
+        return CreateNullable(value).ToOption();
+    }
+
+    public static NonEmptyString CreateNullable(string value)
```

Comment:
I think some comment saying "we treat string with whitespace as null" would be nice

Reply:
Comments will be added also with tests in further PRs.